libs/grok/src/main/java/org/elasticsearch/grok/Grok.java

```diff
@@ -86,11 +86,52 @@ public Grok(Map<String, String> patternBank, String grokPattern) {
         this.patternBank = patternBank;
         this.namedCaptures = namedCaptures;
 
-        this.expression = toRegex(grokPattern);
+        for (Map.Entry<String, String> entry : patternBank.entrySet()) {
+            String name = entry.getKey();
+            String pattern = entry.getValue();
+            forbidCircularReferences(name, new ArrayList<>(), pattern);
+        }
+
+        String expression = toRegex(grokPattern);
         byte[] expressionBytes = expression.getBytes(StandardCharsets.UTF_8);
         this.compiledExpression = new Regex(expressionBytes, 0, expressionBytes.length, Option.DEFAULT, UTF8Encoding.INSTANCE);
     }
 
+    private void forbidCircularReferences(String patternName, List<String> path, String pattern) {
```

Comment:
we should probably add some javadocs here so when anyone picks this up in the future, they will have some context. it is very easy to get into the details of all the curly braces and lose track of why we are doing this.